KAFKA-12276: Add the quorum controller code

cmccabe opened the pull request:
The quorum controller stores metadata in the KIP-500 metadata log, not in Apache ZooKeeper.  Each controller node is a voter in the metadata quorum.  The leader of the quorum is the active controller, which processes write requests.  The followers are standby controllers, which replay the operations written to the log.  If the active controller goes away, a standby controller can take its place.

Like the ZooKeeper-based controller, the quorum controller is based on an event queue backed by a single-threaded executor.  However, unlike the ZK-based controller, the quorum controller can have multiple operations in flight-- it does not need to wait for one operation to be finished before starting another.  Therefore, calls into the QuorumController return CompleteableFuture objects which are completed with either a result or an error when the operation is done.  The QuorumController will also time out operations that have been sitting on the queue too long without being processed.  In this case, the future is completed with a TimeoutException.

The controller uses timeline data structures to store multiple "versions" of its in-memory state simultaneously.  "Read operations" read only committed state, which is slightly older than the most up-to-date in-memory state.  "Write operations" read and write the latest in-memory state.  However, we can not return a successful result for a write operation until its state has been committed to the log.  Therefore, if a client receives an RPC response, it knows that the requested operation has been performed, and can not be undone by a controller failover.

cmccabe committed:
KAFKA-12276: Add the KIP-500 controller code

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+    void remove(int brokerId) {
+        BrokerHeartbeatState broker = brokers.remove(brokerId);
+        if (broker != null) {
+            if (!broker.fenced()) {
+                unfenced.remove(broker);
+            }
+            if (!broker.shuttingDown()) {
+                active.remove(broker);
+            }
```
Should the second check appear within the first `if` as it does below in `touch()`?  And assuming yes, maybe refactor that common logic out into a `private void removeFromActiveAndUnfenced(BrokerHeartbeatState broker)` method?

cmccabe replied:
Thanks... that is a good catch.  Yes, it's a bit more efficient if the statements are nested.  I will refactor this out into a separate function.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+     *
+     * @param brokerId      The broker ID to remove.
+     */
+    void remove(int brokerId) {
```
`public void remove(...)`?

cmccabe replied:
Hmm, are you suggesting that it should be public?  I'd rather not make this public because it's only accessed from within the controller package

rondagostino replied:
> only accessed from within the controller package

Ok, I was just checking.  These are all fine then.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+     *
+     * @return              True if the given broker has a valid session.
+     */
+    boolean hasValidSession(int brokerId) {
```
`boolean hasValidSession(...)`?

cmccabe replied:
Hmm, I'm not sure I understand the question....

rondagostino replied:
Sorry, was asking if it should be public -- but I assume not.  Was just checking.  It's fine now.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+     * @param fenced            True only if the broker is currently fenced.
+     * @param metadataOffset    The latest metadata offset of the broker.
+     */
+    void touch(int brokerId, boolean fenced, long metadataOffset) {
```
`public void touch(...)`?

cmccabe replied:
Hmm, are you suggesting that it should be public?  I'd rather not make this public because it's only accessed from within the controller package

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+        }
+        broker.lastContactNs = time.nanoseconds();
+        broker.metadataOffset = metadataOffset;
+        if (fenced) {
+            broker.shutdownOffset = -1;
```
This seems to imply that it is impossible for a broker to be doing a controlled shutdown and be fenced.  I guess that means any controlled shutdown gets cancelled?  A comment explaining the implications would be helpful.  Actually, from further down in `shouldShutDown()` it appears it can shutdown immediately if it is fenced -- so I think it's about leaders moving away?  Again, a comment would help.

cmccabe replied:
Good question.  A fenced broker will not have leaders, so there should be no leaders to move away.  More specifically, if any fenced broker tries to enter controlled shutdown, it will be shut down immediately.  I'll add a comment.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+            broker = new BrokerHeartbeatState(brokerId);
+            brokers.put(brokerId, broker);
+        } else {
+            if (!broker.fenced()) {
+                unfenced.remove(broker);
+                if (!broker.shuttingDown()) {
+                    active.remove(broker);
+                }
+            }
```
At first I was confused as to why these two operations were necessary, then I realized it is because the instance is mutable and its places in the ordered list and `TreeSet` are going to change.  A comment here would be helpful to make this apparent (I know there is a comment in the list and set declarations, but a reminder here would be helpful nonetheless).

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+        }
+    }
+
+    void beginBrokerShutDown(int brokerId, boolean deferred) {
```
`public void beginBrokerShutDown(...)`?  Javadoc would be helpful, especially to explain what `deferred` is about.  Or, if `private` rather than `public`, at least a comment.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+        if (broker == null) {
+            throw new RuntimeException("Unable to locate broker " + brokerId);
+        }
+        if (!broker.shuttingDown()) {
```
What is supposed to happen if it is already shutting down and this is invoked?  Will it matter if `deferred` is different in the second call?

cmccabe replied:
If it's already shutting down nothing happens

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+        }
+    }
+
+    boolean shouldShutDown(int brokerId) {
```
`public` or `private`?  Same with methods below.

cmccabe replied:
package-private is OK

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+        if (broker == null) {
+            throw new RuntimeException("Unable to locate broker " + brokerId);
+        }
+        if (broker.fenced()) {
+            log.debug("broker {} can shut down immediately because it is fenced.");
+            return true;
+        }
```
Is this the case because fenced implies leadership is already moving away?  If so, a comment to that effect (or some additional wording in the log line) would be helpful.

cmccabe replied:
In general it doesn't make sense to wait for controlled shutdown if the broker is already fenced, because in that case its leaders have already been moved away.  I will add a comment.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+        broker.lastContactNs = time.nanoseconds();
+        broker.metadataOffset = metadataOffset;
+        if (fenced) {
+            broker.shutdownOffset = -1;
```
You seem to sometime use `shutdown` and other times use `shutDown` -- not sure if that is on purpose or there is a lack of consistency?

cmccabe replied:
I wanted to standardize on shutDown.  I will fix the inconsistency.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+                "shutdown state.", brokerId);
+            return false;
+        }
+        long curOffset = lowestActiveOffset();
```
`currLowestActiveOffset` a better name?

cmccabe replied:
I'll switch to `lowestActiveOffset`

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,491 @@
+        return first.metadataOffset;
+    }
+
+    void updateShutdownOffset(int brokerId, long offset) {
```
What's the difference between `beginBrokerShutDown()` and `updateShutdownOffset()`?  Why would the offset at which it can shutdown change?  A comment would be helpful.

rondagostino reviewed (commented):
Just got through a little bit, figured I would provide incremental feedback.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/Controller.java:
```diff
@@ -0,0 +1,180 @@
+     *
+     * @param brokerId      The broker id to decommission.
+     *
+     * @return              A future that is completed successfully hwne the broker is
```
typo hwne

junrao commented on metadata/src/main/java/org/apache/kafka/controller/Controller.java:
```diff
@@ -0,0 +1,180 @@
+     * @param configChanges The changes.
+     * @param validateOnly  True if we should validate the changes but not apply them.
+     *
+     * @return              A future yielding a map from partitions to error results.
```
The returned map is not keyed on partition.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/Controller.java:
```diff
@@ -0,0 +1,180 @@
+     * @param newConfigs    The new configuration maps to apply.
+     * @param validateOnly  True if we should validate the changes but not apply them.
+     *
+     * @return              A future yielding a map from partitions to error results.
```
The returned map is not keyed on partition.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,917 @@
+    private void handleEventEnd(String name, long startProcessingTimeNs) {
+        long endProcessingTime = time.nanoseconds();
+        long deltaNs = endProcessingTime - startProcessingTimeNs;
+        log.debug("Processed {} in {} us", name,
```
Do we also need to update the metric for processing time?

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,917 @@
+    private static final String ACTIVE_CONTROLLER_EXCEPTION_TEXT_PREFIX =
+        "The active controller appears to be node ";
+
+    private NotControllerException newNotControllerException() {
```
Could we move those private methods after all the internal classes?

cmccabe replied:
I normally do group private methods after internal classes, but in this case, it seemed better to keep them together.  Otherwise you'd have to jump around a lot when reading the code.  What do you think?

junrao replied:
Sounds good too.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,917 @@
+         * parameters, for example.  Of course, partial errors are possible for batch
+         * operations.
+         *
+         * @return              A tuple containing a list of records, and an RPC result.
```
The return type is not a tuple.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ControllerPurgatory.java:
```diff
@@ -0,0 +1,108 @@
+            if (offset < lastKey) {
+                throw new RuntimeException("There is already a purgatory event with " +
+                    "offset " + offset + ".  We should not add one with an offset which " +
+                    "is lower than or equal to that.");
```
Could we log lastKey too?

junrao commented on metadata/src/main/java/org/apache/kafka/metalog/LocalLogManager.java:
```diff
@@ -0,0 +1,379 @@
+
+
+/**
+ * The LocalLogManager is a test implementation that relies on the contents of memory.
```
Is this used only for test?

cmccabe replied:
Yes, it is.  I will move it to the test directory.

junrao commented on metadata/src/main/java/org/apache/kafka/metalog/LocalLogManager.java:
```diff
@@ -0,0 +1,379 @@
+    public static class SharedLogData {
+        private final Logger log = LoggerFactory.getLogger(SharedLogData.class);
+        private final HashMap<Integer, LocalLogManager> logManagers = new HashMap<>();
+        private final TreeMap<Long, LocalBatch> batches = new TreeMap<>();
```
Could we add a bit comment explaining logManagers and batches?

junrao commented on metadata/src/main/java/org/apache/kafka/metalog/LocalLogManager.java:
```diff
@@ -0,0 +1,379 @@
+
+    private boolean shutdown = false;
+
+    private long maxReadOffset = Long.MAX_VALUE;
```
Could we add a comment for maxReadOffset? Is it the committed offset?

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+                        }
+                    }
+                    for (ApiMessage message : messages) {
+                        replay(message);
```
We already replay the message when it's first appended to the log and here we replay the same message again after commit. This could temporarily revert the state. For example, the latest (uncommitted) config could be overwritten by a previously committed config.

cmccabe replied:
This code is only executed by the followers.  It is true that the leader already applied these log messages but these nodes are not the leader.

junrao replied:
Sounds good. Could we add a comment to make that clear?

cmccabe replied:
added

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+    }
+
+    ControllerResult<AlterIsrResponseData> alterIsr(AlterIsrRequestData request) {
+        clusterControl.checkBrokerEpoch(request.brokerId(), request.brokerEpoch());
```
This can throw StaleBrokerEpochException. It would be useful for KafkaEventQueue.run() to log the event associated with the exception.

cmccabe replied:
`handleEventException` handles logging exceptions thrown by events.

junrao replied:
It seems that we are logging at the debug level. I am wondering if we should log at WARN as before in ZK based appoach.

```
        if (exception instanceof ApiException) {
            log.debug("{}: failed with {} in {} us", name,
                exception.getClass().getSimpleName(), deltaUs);
            return exception;
        }

```

cmccabe replied:
We can get here just because the user made an invalid RPC, so I don't know if WARN is appropriate.  I'll change it to INFO for now.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+                        setErrorCode(Errors.FENCED_LEADER_EPOCH.code()));
+                    continue;
+                }
+                if (partitionData.currentIsrVersion() != partition.partitionEpoch) {
```
In the ZK case, we use the ZK version to do conditional updates. In Raft, could we associated each partitionState with the offset in the Raft log and use that as partitionEpoch for conditional updates? This way, we don't need to explicitly maintain a separate partitionEpoch field and the epoch is automatically bumped up for any change to the partition record, not just for leader and isr.

cmccabe replied:
Yes, I think that could work for partition epoch.  Let's do that once we have the initial code in, though...

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+     * Otherwise, this is -1.  This variable must be modified only from the controller
+     * thread, but it can be read from other threads.
+     */
+    private volatile long curClaimEpoch;
```
I thought the raft leader epoch is an int since we store only int as leader epoch in the log?

cmccabe replied:
It is an int and will be in 2.8, but I think that's a mistake (as discussed in the mailing list) and we should plan to make this 64 bit in the near future to avoid overflow issues.

So the controller code treats it as a long... cc @hachikuji

junrao replied:
I thought the issue was that if changing leaderEpoch to Long requires a log format change for user data, it has significant performance impact such as down conversion.

junrao replied:
It seems that using long for leaderEpoch in the log requires a bigger discussion. Could we use int for now?

cmccabe replied:
I do think we should harmonize this, but I think it would be better to do that when we get rid of MetaLogShim. We've had a plan to get rid of the shim layer for a while but we just didn't have time to do it this week. So let's plan to do it then, if that makes sense

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+                    newLeader = partition.leader;
+                    newLeaderEpoch = partition.leaderEpoch;
+                } else {
+                    newLeader = partition.chooseNewLeader(newIsr, false);
```
Currently, the follower never removes the leader out of ISR. So, perhaps we should just throw an exception if this is not the case.

cmccabe replied:
ok.  We can make this an invalid request then.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+        }
+        BrokerRegistration existing = brokerRegistrations.get(brokerId);
+        if (existing == null) {
+            return new ControllerResult<>(new ArrayList<>(), null);
```
This is in response to a heartBeat request. So, it should generate a response in ControllerResult?

cmccabe replied:
Hmm, this comment shows up for me as being left in the `decommissionBroker` function, which is called in response to the decomission broker RPC, not the heartbeat RPC.  Maybe this is a Github UI issue?  Did you mean to leave this comment for a different function?

junrao replied:
Sorry, I meant `decommissionBroker`. It seems that decommissionBroker request should we a response too, right?

cmccabe replied:
Note: this has been changed to unregisterBroker as per the mailing list discussion.

It's OK to have a future that returns void.  That just means you either get success, or an error (there is no other result).  Which is consistent with UnregisterBrokerResponse.json, etc.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+                this.resultAndOffset = new ControllerResultAndOffset<>(offset,
+                    result.records(), result.response());
+                for (ApiMessageAndVersion message : result.records()) {
+                    replay(message.message());
```
Some of the replay (e.g. UNREGISTER_BROKER_RECORD) could throw exceptions. We probably need to turn the exception into an error response. Are we handling that already?

cmccabe replied:
If an exception is thrown here we will end up in `handleEventException`.  Since the exception won't be a subclass of `ApiException` we will resign as controller and return an `UnknownServerException`.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+        return new ControllerResult<>(records, response);
+    }
+
+    void removeFromIsr(int brokerId, List<ApiMessageAndVersion> records) {
```
We probably should name this sth like removeFromIsrAndMaybeChooseLeader.

cmccabe replied:
I changed it to `removeFromIsrAndLeaderships`

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+                }
+            }
+            if (unclean && replicas.length > 0) {
+                return replicas[0];
```
We need to choose at least a live replica.

cmccabe replied:
good point, will fix

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ConfigurationControlManager.java:
```diff
@@ -0,0 +1,367 @@
+    static ApiError checkConfigResource(ConfigResource configResource) {
+        switch (configResource.type()) {
+            case BROKER_LOGGER:
+                // We do not handle resources of type BROKER_LOGGER in
```
Is that temporary?

cmccabe replied:
This has to be handled by the individual brokers.  It's not persisted anywhere currently (currently it is not stored in ZK I believe)

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+        int chooseNewLeader(int[] newIsr, boolean unclean) {
+            for (int i = 0; i < replicas.length; i++) {
+                int replica = replicas[i];
+                if (Replicas.contains(newIsr, replica)) {
```
In the ZK based code, we also take live brokers into consideration when selecting a new leader.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+                    " existed in isrMembers, but not in the partitions map.");
+            }
+            int[] isrWithoutCurLeader = Replicas.copyWithout(partition.isr, brokerId);
+            int newLeader = partition.chooseNewLeader(isrWithoutCurLeader, false);
```
Hmm, not all partitions with ISR containing the shutting down need to change the leader.

cmccabe replied:
In this function we are iterating only over the partitions that the given broker is a leader for.

( We obtained the iterator from `BrokersToIsrs#iterator(brokerId=brokerId, leadersOnly=true)` )

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+        }
+        int brokerId = request.brokerId();
+        BrokerRegistration registration = brokerRegistrations.get(brokerId);
+        if (registration == null) {
```
We already did this check and the one below in the caller through `clusterControl.checkBrokerEpoch`.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+        if (isFenced) {
+            if (request.wantShutDown()) {
+                // If the broker is fenced, and requests a shutdown, do it immediately.
+                isFenced = true;
```
This seems unnecessary.

cmccabe replied:
ok

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+                if (heartbeatManager.shouldShutDown(request.brokerId())) {
+                    isFenced = true;
+                    shouldShutdown = true;
+                    records.add(new ApiMessageAndVersion(new FenceBrokerRecord().
```
If we want to log the shutting down the broker, it seems it's more consistent if we always log it. Now, it seems we log it only when leaders need to be moved.

cmccabe replied:
I will improve the logging a bit here.  I agree that we should log when a broker is told it can shut down or be (un) fenced, since those are major events.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+                updateShutdownOffset = true;
+            }
+            if (result.response().isFenced()) {
+                replicationControl.removeFromIsr(request.brokerId(), records);
```
It seems that we if the request wants to shut down, we should always remove the shutting down broker from isr, just like moving the leader off the shutting down broker?

cmccabe replied:
It seemed safer to leave it in the ISR until it's ready to shut down for good.  Also, if we take it out, it might just get re-added if it catches up... ?

junrao replied:
Well, in ZK based approach, in response to a controlled shutdown, the controller changes ISR and the leader. Here, it seems that `result.response().isFenced()` is not always true if the broker heartbeat indicates the intention to shut down?

cmccabe replied:
Hmm... We only fence the broker once controlled shutdown has completed.  If we fenced it immediately that would be disruptive to clients since they wouldn't be able to continue fetching from it until the leaderships have moved. Basically immediate fencing is the NON-controlled shutdown path....

junrao reviewed (commented):
@cmccabe : Thanks for the PR. Partially through the PR. A few comments below.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+            throw new RuntimeException(String.format("Unable to replay %s: no broker " +
+                "registration with that epoch found", record.toString()));
+        } else {
+            brokerRegistrations.put(brokerId, registration.cloneWithFencing(false));
```
When a broker is unfenced, some of the partitions without leader could have a new leader now. So, it seems that we need to trigger a leader election here.

cmccabe replied:
Good point. The appropriate place to handle this will be in the broker heartbeat handling code, since that is where the active controller unfences brokers.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+            List<ApiMessageAndVersion> records = new ArrayList<>();
+            boolean movingLeadersForShutDown = false;
+            if (request.wantShutDown()) {
+                records = replicationControl.removeLeaderships(request.brokerId());
```
This call generates a leader change record before the following fencedBroker record. Ordering wise, it seems that we should record the fencedBroker first. 

Also, I am wondering what's the best place to trigger leader election and removing from ISR due to fenced broker. There are multiple cases when a broker can be fenced (e.g. broker controlled shutdown, broker fenced due to no heartbeat). Instead of of doing leader election and isr removal in all those cases, another option is to tigger them in a single place when a fencedBroker record is replayed.

cmccabe replied:
> This call generates a leader change record before the following fencedBroker record. Ordering wise, it seems that we should record the fencedBroker first.

Hmm... it should be OK to remove the leaderships first.

Kip-500 Controlled shutdown works this way, for example... the shutting-down broker is not actually fenced at all until all the other brokers have removed it as a leader.

Also, wouldn't it be a bit weird to be in a state where a broker is still marked as the leader for some partition, but doesn't show up in the list of brokers given in the MetadataResponse?  That would happen if we put the fencing record first.  I don't think clients or brokers would handle this well.

> Also, I am wondering what's the best place to trigger leader election and removing from ISR due to fenced broker. There are multiple cases when a broker can be fenced (e.g. broker controlled shutdown, broker fenced due to no heartbeat). Instead of of doing leader election and isr removal in all those cases, another option is to tigger them in a single place when a fencedBroker record is replayed.

Replaying a record cannot trigger the creation of any additional records.  This would not work since the standby controllers can't create records, after all... only the active controller.

I have created a `handleBrokerFenced` function in `ReplicationControlManager` which does most of what needs to be done for fencing, though... aside from creating the actual fencing record.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+            }
+            int[] isrWithoutCurLeader = Replicas.copyWithout(partition.isr, brokerId);
+            int newLeader = partition.chooseNewLeader(isrWithoutCurLeader, false);
+            records.add(new ApiMessageAndVersion(new IsrChangeRecord().
```
Since we use IsrChangeRecord for changing the leader too, could we name it to sth more general?

cmccabe replied:
I have renamed it to `PartitionChangeRecord`

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+    public CompletableFuture<Void> decommissionBroker(int brokerId) {
+        return appendWriteEvent("decommissionBroker", () -> {
+            ControllerResult<Void> result = clusterControl.decommissionBroker(brokerId);
+            replicationControl.removeFromIsr(brokerId, result.records());
```
Should we also trigger leader elections here? Also, should we allow broker decommission when it still has replicas?

cmccabe replied:
> Should we also trigger leader elections here? 

Good catch.  Yes, this should be moving leaders.  I fixed this in the latest version of the PR.

> Also, should we allow broker decommission when it still has replicas?

Yes, this is allowed.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+            ControllerResult<Set<Integer>> result =
+                clusterControl.maybeFenceLeastRecentlyContacted();
+            for (int brokerId : result.response()) {
+                replicationControl.removeFromIsr(brokerId, result.records());
```
This should tigger leader election too, right?

cmccabe replied:
I restructured this code a bit.  But yes, it does move leaders if needed (it's clearer in the new version I think)

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+                if (request.wantShutDown()) {
+                    heartbeatManager.beginBrokerShutDown(request.brokerId(), movingLeadersForShutDown);
+                }
+                if (heartbeatManager.shouldShutDown(request.brokerId())) {
```
Is the intention of `shouldShutDown` to wait until the metadata of the shutting down broker is received by other brokers? If so, not sure why `beginBrokerShutDown` sets `broker.shutDownOffset` to either MAX_LONG or 0.

Also, if `shouldShutDown` returns false, when do we get another opportunity to mark the shutting down broker as fenced?

cmccabe replied:
This got refactored, hopefully the new version is clearer.  The new version avoids the MAX_LONG / 0 hack and other ugliness that was in the initial version.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+                shouldShutdown = request.wantShutDown();
+            } else {
+                if (request.wantShutDown()) {
+                    heartbeatManager.beginBrokerShutDown(request.brokerId(), movingLeadersForShutDown);
```
In the ZK based logic, on receiving a controlled shutdown request, the controller tries to move the leaders off the broker. If this is successful, the controller sends a successful return for the broker to proceed with shutdown. Here, it seems that the controller will initially return shouldShutDown as false to the broker if there are leaders moved off the broker and require the broker to heartbeat again to be able to shut down.

cmccabe replied:
that is correct.  the broker must send another heartbeat before it is allowed to shut down.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,519 @@
+                                      ReplicaPlacementPolicy policy) {
+        Iterator<UsableBroker> iterator = new UsableBrokerIterator(
+            unfenced.iterator(), idToRack);
+        return policy.createPlacement(numPartitions, numReplicas, iterator);
```
I thought we want to allow topics to be created even when there is not enough live brokers now?

cmccabe replied:
good point.  Since we don't have much time for 2.8, I will add a TODO for now.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+                }
+                // If there are operations in the purgatory, we want to wait for the latest
+                // one to complete before returning our result to the user.
+                this.resultAndOffset = new ControllerResultAndOffset<>(maybeOffset.get(),
```
Do we need to use `this`? Ditto below.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+        CompletableFuture<ElectLeadersResponseData> future = new CompletableFuture<>();
+        appendWriteEvent("electLeaders", request.timeoutMs(),
+            () -> replicationControl.electLeaders(request));
+        return future;
```
Hmm, should we return the future from `appendWriteEvent` ?

cmccabe replied:
good catch.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,456 @@
+        }
+
+        if (existing == null) {
+            heartbeatManager.touch(brokerId, true, -1);
```
Sometimes, we update the in-memory state after the record is appended to the log. Here, it seems that we do the reverse. Should we make that consistent?

cmccabe replied:
The heartbeat manager is special because it stores soft state which is not in the metadata log (when each broker last heartbeated, for example).

junrao commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,519 @@
+     * Move a broker into shutdown state.
+     *
+     * @param brokerId      The broker id.
+     * @param deferred      If this is true, the broker can shut down immediately.
```
It seems the broker can shut down immediately if this is false?

cmccabe replied:
this got refactored. hopefully it is clearer now.  yes, a broker can shut down immediately in some cases

junrao commented on metadata/src/main/java/org/apache/kafka/controller/BrokersToIsrs.java:
```diff
@@ -0,0 +1,308 @@
+        for (int i = 0; i < partitions.length; i++) {
+            int cur = partitions[i];
+            if (cur == target) {
+                newPartitions[i] = isLeader ? partition | LEADER_FLAG : partition;
```
Could we just update `partitions` in place?

cmccabe replied:
since this is stored in a `TimelineHashMap` it must be treated as immutable.  We can't modify the past.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+        }
+
+        @Override
+        public void handleNewLeader(MetaLogLeader newLeader) {
```
What triggers this on a hard controller failure?

cmccabe replied:
this comes out of the Raft layer and is invoked when the Raft election has succeeded and produced a new leader node.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+                    log.info("Becoming active at controller epoch {}.", newEpoch);
+                    curClaimEpoch = newEpoch;
+                    writeOffset = lastCommittedOffset;
+                    clusterControl.activate();
```
Should we trigger the logic for leader election/isr removal since there could be unhandled fencedBroker records when the new controller takes over?

cmccabe replied:
the standby controller must replay all the committed records before becoming active.  so, there is no unfinished work to be done at this point.

junrao replied:
Are all records generated in a single ControllerWriteEvent appended to the metadata log atomically?

cmccabe replied:
RIght now the answer is yes. Eventually we plan on supporting multiple batches.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/Controller.java:
```diff
@@ -0,0 +1,180 @@
+     * @return              A future that is completed successfully hwne the broker is
+     *                      decommissioned, or if it is not registered in the first place.
+     */
+    CompletableFuture<Void> decommissionBroker(int brokerId);
```
Should this be changed to unregisterBroker to match the KIP?

junrao commented on metadata/src/main/java/org/apache/kafka/controller/Controller.java:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * Blocks until we have shut down and freed all resources.
+     */
+    void close() throws InterruptedException;
```
Should we call this sth like waitForShutdownComplete to match `beginShutdown`?

cmccabe replied:
The name close comes from `AutoCloseable`, which makes some of the tests nicer to write (because we can use the Java try-with-resources syntax).

junrao commented on metadata/src/main/java/org/apache/kafka/controller/LeaderAndIsr.java:
```diff
@@ -0,0 +1,51 @@
+    private final int leaderId;
+    private final int leaderEpoch;
+    private final List<Integer> isr;
+    private final int currentZkVersion;
```
Is this still needed with Raft metadata?

junrao commented on metadata/src/main/java/org/apache/kafka/controller/LeaderAndIsr.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.util.List;
+
+public class LeaderAndIsr {
```
This class seems unused.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/PartitionLeaderElectionResult.java:
```diff
@@ -0,0 +1,38 @@
+
+import org.apache.kafka.common.requests.ApiError;
+
+public class PartitionLeaderElectionResult {
```
This class seems unused.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicaPlacementPolicy.java:
```diff
@@ -0,0 +1,46 @@
+
+
+/**
+ * Represents a deferred event in the controller purgatory.
```
Inaccurate comment.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+import static org.apache.kafka.common.config.ConfigResource.Type.TOPIC;
+
+
+public class ReplicationControlManager {
```
Could we add some comment for this class?

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+                prefix = ", ";
+            }
+            if (!Arrays.equals(isr, prev.isr)) {
+                builder.append(prefix).append("isr=").append(Arrays.toString(isr));
```
It might be useful to log both the old and the new value.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+        PartitionControlInfo newPartInfo = new PartitionControlInfo(record);
+        PartitionControlInfo prevPartInfo = topicInfo.parts.get(record.partitionId());
+        if (prevPartInfo == null) {
+            log.info("Created partition {}:{} with {}.", record.topicId(),
```
Are we deprecating the state-change log and the controller log that we had before?

cmccabe replied:
I don't think the state change log can scale to the number of partitions we need.  It gets too verbose.  Also, this information is available in the metadata log itself.

junrao reviewed (commented):
A few more comments.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,920 @@
+    public CompletableFuture<BrokerRegistrationReply>
+            registerBroker(BrokerRegistrationRequestData request) {
+        return appendWriteEvent("registerBroker", () -> {
+            ControllerResult<BrokerRegistrationReply> result = clusterControl.
+                registerBroker(request, writeOffset + 1, featureControl.
+                    finalizedFeaturesAndEpoch(Long.MAX_VALUE));
```
I believe after this we need to invoke something to cover the case where a topic has this broker as its only ISR member:

```
replicationControl.maybeMakeLeader(request.brokerId(), result.records());
```

The implementation might look like this:

```
    void maybeMakeLeader(int brokerId, List<ApiMessageAndVersion> records) {
        Iterator<TopicPartition> iterator = brokersToIsrs.iterator(brokerId, false);
        while (iterator.hasNext()) {
            TopicPartition topicPartition = iterator.next();
            TopicControlInfo topic = topics.get(topicPartition.topicId());
            if (topic == null) {
                throw new RuntimeException("Topic ID " + topicPartition.topicId() + " existed in " +
                    "isrMembers, but not in the topics map.");
            }
            PartitionControlInfo partition = topic.parts.get(topicPartition.partitionId());
            if (partition == null) {
                throw new RuntimeException("Partition " + topicPartition +
                    " existed in isrMembers, but not in the partitions map.");
            }
            // make it the leader if it is the only ISR member
            if (partition.isr.length == 1 && partition.isr[0] == brokerId && partition.leader != brokerId) {
                int newLeader = brokerId;
                int newLeaderEpoch = partition.leaderEpoch + 1;
                IsrChangeRecord message = new IsrChangeRecord().
                    setPartitionId(topicPartition.partitionId()).
                    setTopicId(topic.id).
                    setIsr(Replicas.toList(partition.isr)).
                    setLeader(newLeader).
                    setLeaderEpoch(newLeaderEpoch).
                    setPartitionEpoch(partition.partitionEpoch + 1);
                records.add(new ApiMessageAndVersion(message, (short) 0));
            }
        }
    }
```

cmccabe replied:
Thanks, @rondagostino .  In the latest version, I do not remove brokers from singleton ISRs.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,712 @@
+                throw new RuntimeException("Partition " + topicPartition +
+                    " existed in isrMembers, but not in the partitions map.");
+            }
+            int[] newIsr = Replicas.copyWithout(partition.isr, brokerId);
+            int newLeader, newLeaderEpoch;
+            if (partition.leader == brokerId) {
+                newLeader = partition.chooseNewLeader(newIsr, false);
+                newLeaderEpoch = partition.leaderEpoch + 1;
+            } else {
+                newLeader = partition.leader;
+                newLeaderEpoch = partition.leaderEpoch;
+            }
+            records.add(new ApiMessageAndVersion(new IsrChangeRecord().
+                setPartitionId(topicPartition.partitionId()).
+                setTopicId(topic.id).
+                setIsr(Replicas.toList(newIsr)).
+                setLeader(newLeader).
+                setLeaderEpoch(newLeaderEpoch).
+                setPartitionEpoch(partition.partitionEpoch + 1), (short) 0));
```
I believe we should surround this section of code with the following to be sure we never drop the last ISR member:

```
            // never remove the last ISR member
            if (partition.isr.length > 1) {
                int[] newIsr = ...
                etc...
            }
```

rondagostino reviewed (commented):
Commented in a couple of places where I think we need to deal with the case where the partition's ISR list has just a single member.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,916 @@
+            }
+            if (controllerMetrics == null) {
+                controllerMetrics = (ControllerMetrics) Class.forName(
+                    "org.apache.kafka.controller.MockControllerMetrics").getConstructor().newInstance();
```
`MockControllerMetrics` is a test class?

rondagostino replied:
Maybe it would be better to check for null and exit out if it is unset -- otherwise we see this, which is not ideal:

```
Fatal error during controller startup. Prepare to shutdown (kafka.server.ControllerServer)
java.lang.ClassNotFoundException: org.apache.kafka.controller.MockControllerMetrics
```

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumControllerMetrics.java:
```diff
@@ -0,0 +1,61 @@
+    private final static MetricName ACTIVE_CONTROLLER_COUNT = new MetricName(
+        "kafka.controller", "KafkaController", "ActiveControllerCount", null);
+    private final static MetricName EVENT_QUEUE_TIME_MS = new MetricName(
+        "kafka.controller", "ControllerEventManager", "EventQueueTimeMs", null);
```
In the old controller, EventQueueTimeMs is used to measure the amount of time an event is sitting in the queue before being processed. There is a separate timer metric per controller state that measures the processing time per event type.

cmccabe replied:
Good point. I will fix it so that EventQueueTimeMs has its original meaning.

For now, I have added a metric called EventQueueProcessingTimeMs which deals with processing time.  I do want to do the per-state tracking but I don't think we have time right now

junrao commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,585 @@
+                    } else {
+                        log.info("Unfenced broker {} has requested and been granted an " +
+                            "immediate shutdown.", brokerId);
+                        return new BrokerControlStates(currentState, SHUTDOWN_NOW);
```
If a broker wants to shut down and is only included in ISR, we still want to remove the broker from ISR before allowing it to shutdown. Otherwise, a new published record needs to wait for the session timeout before it can be committed.

cmccabe replied:
OK, that makes sense.  I will remove it from all non-singleton ISRs as well as removing it from all leaderships.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,585 @@
+                return new BrokerControlStates(currentState, FENCED);
+
+            case UNFENCED:
+                if (request.wantShutDown()) {
```
What about the case when request.wantFence() is true?

cmccabe replied:
Good question. The broker doesn't currently request fencing once it is unfenced.  But for completeness, it is simple to support this and it makes the code more intuitive, so I'll add it.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,900 @@
+                    handleBrokerUnfenced(brokerId, brokerEpoch, records);
+                    break;
+                case CONTROLLED_SHUTDOWN:
+                    removeLeaderships(brokerId, records);
```
(1) In ZK-based approach, we do leader election a bit differently for controlled shutdown. If we can't select a leader from the remaining ISR, we just leave the current leader as it is. This gives the shutting down broker a chance to retry controlled shutdown until the timeout.

(2) In ZK-based approach, we also remove the broker from isr for other partitions whose leader is not on the shutting down broker.

> It seemed safer to leave it in the ISR until it's ready to shut down for good. Also, if we take it out, it might just get re-added if it catches up... ?

That's true and is an existing problem. One way to address this is to include partitionEpoch in the follower fetch request. The leader could then reject a follower request if the partitionEpoch doesn't match. This can be done in a followup PR.

cmccabe replied:
As I mentioned above, I changed it so that it now removes the broker from all non-singleton ISRs, as well as removing it from leaderships.

It seems like the remaining behavioral difference is that the new code will, if no other leader can be chosen, set the leader to -1 (offline).  If we don't do this, controlled shutdown easily gets stuck if there are any partitions with replication factor = 1.  Maybe we can tune this a bit later?

I like the idea of including the partition epoch in the follower fetch request.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,900 @@
+    }
+
+    ControllerResult<ElectLeadersResponseData> electLeaders(ElectLeadersRequestData request) {
+        boolean unclean = electionIsUnclean(request.electionType());
```
Could we add a TODO for handling the preferred leader election?

cmccabe replied:
Hmm, I thought this already handles preferred leader election (there are only two options, PREFERRED and UNCLEAN, so far...)

junrao replied:
I think we need to handle preferred leader election in a special way. For example, if the assigned replicas are 1,2,3, isr is 2,3 and the current leader is 3, when doing preferred leader election, we want to keep the leader as 3 instead of changing it to 2.

junrao replied:
> > I think we need to handle preferred leader election in a special way. For example, if the assigned replicas are 1,2,3, isr is 2,3 and the current leader is 3, when doing preferred leader election, we want to keep the leader as 3 instead of changing it to 2.


> 
> Hmm, wouldn't we want to switch the leader to 2 in that case, since 2 is more preferred?

Well, currently the contract is just that if every broker picks the preferred replica (i.e. 1st replica), the leaders will be balanced among brokers. If not, all other replicas are equivalent. Moving leaders among non-preferred replicas just creates churns without benefiting the balance.

cmccabe replied:
ack.  I fixed this

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,359 @@
+        return new ControllerResult<>(records, new BrokerRegistrationReply(brokerEpoch));
+    }
+
+    public ControllerResult<Void> decommissionBroker(int brokerId) {
```
Should this be named unregisterBroker?

cmccabe replied:
Good catch. This has been superseded by `ReplicationControlManager#unregsiterBroker`.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,900 @@
+            }
+            // TODO: if this partition is configured for unclean leader election,
+            // check the replica set rather than the ISR.
+            if (Replicas.contains(partition.isr, brokerId)) {
```
Is this check already implied since we are iterating `brokersToIsrs`?

cmccabe replied:
We're iterating over the partitions with no leader, which may or may not have the newly activated broker in their ISR.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,585 @@
+                        "since the lowest active offset {} is now greater than the " +
+                        "broker's controlled shutdown offset {}.", brokerId,
+                        lowestActiveOffset, broker.controlledShutDownOffset);
+                    return new BrokerControlStates(currentState, SHUTDOWN_NOW);
```
Is there any benefit/enough to only allow the broker to shutdown when all brokers have caught up to controlledShutDownOffset? The ZK-based criteria is that a broker is allowed to shut down as long as all leaders have been moved off the shutting down broker.

cmccabe replied:
Basically it lets us know that the other brokers have successfully taken over as leader (where needed) which avoids having a period of unavailability, ideally

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,900 @@
+     * Generate the appropriate records to handle a broker being unregistered.
+     *
+     * First, we remove this broker from any non-singleton ISR. Then we generate a
+     * FenceBrokerRecord.
```
We are generating an UnregisterBrokerRecord.

junrao commented on metadata/src/main/java/org/apache/kafka/metadata/BrokerHeartbeatReply.java:
```diff
@@ -31,16 +31,23 @@
     private final boolean isFenced;
 
+    /**
+     * True if the heartbeat reply should tell the broker that it should shut down.
```
Hmm, the comment seems to be the same as shouldShutDown.

cmccabe replied:
fixed

junrao commented on metadata/src/test/java/org/apache/kafka/controller/BrokerHeartbeatManagerTest.java:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    @Test
+    public void testMaybeFenceLeastRecentlyContacted() {
```
testFindStaleBrokers ?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ClientQuotaControlManagerTest.java:
```diff
@@ -0,0 +1,238 @@
+        assertInvalidEntity(manager, new ClientQuotaEntity(Collections.emptyMap()));
+    }
+
+    void assertInvalidEntity(ClientQuotaControlManager manager, ClientQuotaEntity entity) {
```
Could we make it private?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Test
+    public void testDecommission() throws Exception {
```
testUnregister?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -0,0 +1,147 @@
+
+    @ParameterizedTest
+    @ValueSource(ints = {3, 10})
+    public void testChooseRandomRegistered(int numUsableBrokers) throws Exception {
```
testPlaceReplicas ?

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. A few more comments.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ConfigurationControlManager.java:
```diff
@@ -0,0 +1,367 @@
+                }
+                return ApiError.NONE;
+            case TOPIC:
+                if (!configResource.name().isEmpty()) {
```
An empty topic name currently results in an INVALID_REQUEST error.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ConfigurationControlManager.java:
```diff
@@ -0,0 +1,367 @@
+                return new ApiError(Errors.INVALID_REQUEST, "Unsupported " +
+                    "configuration resource type BROKER_LOGGER ");
+            case BROKER:
+                if (!configResource.name().isEmpty()) {
```
An empty broker currently results in an INVALID_REQUEST error.

cmccabe replied:
A Resource with type BROKER and an empty name represents a cluster configuration that applies to all brokers.  I'll add a comment

cmccabe committed:
Address review comments

junrao commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -0,0 +1,597 @@
+
+            case UNFENCED:
+                if (request.wantFence()) {
+                    if (request.wantShutDown()) {
```
Do we allow a heartbeat request to set both the fence and wantShutDown flag?

cmccabe replied:
yes, it can set both

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,875 @@
+                    handleBrokerUnfenced(brokerId, brokerEpoch, records);
+                    break;
+                case CONTROLLED_SHUTDOWN:
+                    handleNodeDeactivated(brokerId, records);
```
> It seems like the remaining behavioral difference is that the new code will, if no other leader can be chosen, set the leader to -1 (offline). If we don't do this, controlled shutdown easily gets stuck if there are any partitions with replication factor = 1. Maybe we can tune this a bit later?

It's fine to revisit that later. The tradeoff is that if we wait, it slightly increases the probability of availability since another replica could join isr.

cmccabe replied:
This is resolved in the latest version of the code, where we disable metadata updates on the shutting down broker before starting controlled shutdown, and bump the leader epoch of all partitions.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -0,0 +1,136 @@
+            return new ApiError(Errors.INVALID_UPDATE_VERSION,
+                "The upper value for the new range cannot be less than 1.");
+        }
+        VersionRange localRange = supportedFeatures.get(featureName);
```
This can be revisited later. When finalizing a feature, should be consider other controller's supported features too?

cmccabe replied:
Hmm... right now, we don't have a good way of finding out what features the other controllers support.  Maybe we will have to think more about this when we support rolling upgrade in kip-500.

junrao replied:
Yes, that's the problem. From a consistency perspective, it seems that we should use the supported features from either all controller nodes or none.

cmccabe replied:
let's revisit this after 2.8

junrao replied:
Do we have a jira to track this?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/MockControllerMetrics.java:
```diff
@@ -0,0 +1,47 @@
+package org.apache.kafka.controller;
+
+
+public final class MockControllerMetrics implements ControllerMetrics {
```
This class seems never used?

cmccabe replied:
It's used in unit tests

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -0,0 +1,925 @@
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+
+public final class QuorumController implements Controller {
```
Could we add some comments for this class?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTestEnv.java:
```diff
@@ -0,0 +1,89 @@
+            QuorumController activeController = null;
+            for (QuorumController controller : controllers) {
+                long curEpoch = controller.curClaimEpoch();
+                if (curEpoch != -1) {
```
To make this more intuitive, perhaps we could add a method isActive in QuorumController?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+                            new CreatableTopic().setName("foo").setNumPartitions(1).
+                                setReplicationFactor((short) 1)).iterator()));
+                assertEquals(Errors.INVALID_REPLICATION_FACTOR.code(), active.createTopics(
+                    createTopicsRequestData).get().topics().find("foo").errorCode());
```
Hmm, why is a replication factor of 1 invalid?

cmccabe replied:
There are no unfenced brokers (as you mentioned earlier, we should change this so that it places on the fenced broker).  I will add a TODO

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. Just a few more comments.

cmccabe committed:
Address review comments

cmccabe commented:
> I think we need to handle preferred leader election in a special way. For example, if the assigned replicas are 1,2,3, isr is 2,3 and the current leader is 3, when doing preferred leader election, we want to keep the leader as 3 instead of changing it to 2.

Hmm, wouldn't we want to switch the leader to 2 in that case, since 2 is more preferred?

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -0,0 +1,341 @@
+import java.util.concurrent.CompletableFuture;
+
+
+public class ClusterControlManager {
```
Could we add some comments to this class?

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,875 @@
+                throw new RuntimeException("Partition " + topicPartition +
+                    " existed in isrMembers, but not in the partitions map.");
+            }
+            int[] newIsr = Replicas.copyWithout(partition.isr, brokerId);
```
As Jason pointed out, in ZK based approach, the controller bumps up the leader epoch for removing replica from ISR too.

Also, since the broker is no longer receiving the leaderAndIsr requests, we need some logic for the broker to ignore the new partition record (for follower fetching) once it starts the controlled shutdown process.

cmccabe replied:
I changed this so that the leader epoch is bumped if and only if there is a leader present in the PartitionChange record.  (It is possible to bump the epoch without changing the leader by including the same leader again in the record.)

We now use this during controlled shutdown to unconditionally bump the leader epochs.  Otherwise, we only bump the leader epochs if the leader changed.

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. A few more comments.

cmccabe committed:
Some partition management improvements

Make epoch bumps implicit in PartitionChangeRecord

Always bump the leader epoch of a partition led by a broker in controlled shutdown

cmccabe committed:
Shut down the broker metadata listener during controlled shutdown

cmccabe committed:
fix leader election

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,894 @@
+            int[] newIsr = (record.isr() == null) ? isr : Replicas.toArray(record.isr());
+            int newLeader;
+            int newLeaderEpoch;
+            if (record.leader() == Integer.MIN_VALUE) {
```
Hmm, does Integer.MIN_VALUE have any special meaning? If so, could we use a more intuitive constant?

cmccabe replied:
No special meaning.  It's just a constant that can't be a valid leader.  We could use -2 if that seems nicer.

junrao replied:
Ok, maybe the check can be `record.leader()< -1`?

cmccabe replied:
I added a constant.  I think it looks a little nicer...

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,894 @@
+            int[] newIsr = Replicas.copyWithout(partition.isr, brokerId);
+            if (newIsr.length == 0) {
+                // We don't want to shrink the ISR to size 0. So, leave the node in the ISR.
+                if (alwaysBumpLeaderEpoch || record.leader() != -1) {
```
Hmm, if the leader is already -1 and we can't change ISR, there is no need to generate a new PartitionChangeRecord just to bump up the leader epoch. It won't help controlled shutdown since there is already no leader.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,894 @@
+     *                              always increments the leader epoch.
+     * @param records               The record list to append to.
+     */
+    void handleNodeDeactivated(int brokerId, boolean alwaysBumpLeaderEpoch,
```
Currently, for controller initiated ISR change (controlled shutdown or hard failure), we always bump up the leader epoch. Also, the name alwaysBumpLeaderEpoch is a bit weird since the code in handleNodeDeactivated() doesn't directly bump up leader epoch.

cmccabe replied:
Hmm... I don't completely understand why we would want to bump the leader epoch when the controller removes a non-leader broker B but not when an AlterIsrRequest removes a non-leader broker B from the ISR.  It seems like we should either bump in both scenarios or neither.

Is the fact that we bump in the first scenario just an artifact of the fact that otherwise we could not send out a leader and isr request that had a new epoch and thereby caused a  change?

cmccabe replied:
As per our discussion outside github, let's just use the old behavior for now.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,894 @@
+                newLeaderEpoch = leaderEpoch;
+            } else {
+                newLeader = record.leader();
+                newLeaderEpoch = leaderEpoch + 1;
```
Currently, for leader initiated AlterIsr request, the controller doesn't bump up the leader epoch. If we change that, it will slightly increase unavailability since all clients have to refresh the metadata in this case.

cmccabe replied:
Hmm... ReplicationControlManager should not allow this to happen during an alter isr request.  There is some code that checks if the alter isr request is attempting to remove the current leader from the isr, and returns an error if so.  So the leader should not be changed by an alter isr request and therefore the leader epoch will not be.

```
                if (!Replicas.contains(newIsr, partition.leader)) {
                    // An alterIsr request can't remove the current leader.
                    responseTopicData.partitions().add(new AlterIsrResponseData.PartitionData().
                        setPartitionIndex(partitionData.partitionIndex()).
                        setErrorCode(Errors.INVALID_REQUEST.code()));
                    continue;
                }
```

junrao replied:
Yes, the alterIsr doesn't change leader, but generates a PartitionChangeRecord. On replaying this record, the code following code bumps on leaderEpoch?

`        PartitionControlInfo newPartitionInfo = prevPartitionInfo.merge(record);`

cmccabe replied:
merge only bumps the epoch if the leader was set.

```
        PartitionControlInfo merge(PartitionChangeRecord record) {
            int[] newIsr = (record.isr() == null) ? isr : Replicas.toArray(record.isr());
            int newLeader;
            int newLeaderEpoch;
            if (record.leader() == NO_LEADER_CHANGE) {
                newLeader = leader;
                newLeaderEpoch = leaderEpoch;
            } else {
                newLeader = record.leader();
                newLeaderEpoch = leaderEpoch + 1;
            }
            return new PartitionControlInfo(replicas,
                newIsr,
                removingReplicas,
                addingReplicas,
                newLeader,
                newLeaderEpoch,
                partitionEpoch + 1);
        }
```

junrao commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -397,6 +397,9 @@ class BrokerServer(
       if (config.controlledShutdownEnable) {
+        // Shut down the broker metadata listener, so that we don't get added to any
+        // more ISRs.
+        brokerMetadataListener.beginShutdown()
```
If we do this, does `brokerMetadataListener.close() `still need to call `beginShutdown()`.

cmccabe replied:
It does need to be because there are some paths through the code that don't go through here.  In general calling `beginShutdown` or `close` multiple times is harmless-- only the first time has an effect.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,894 @@
+            // nothing to do.
+            return ApiError.NONE;
+        }
+        if (partitionInfo.leader != -1 && newLeader != partitionInfo.preferredReplica()) {
```
Hmm, it seems that we should only do `newLeader != partitionInfo.preferredReplica()` if this is a preferred leader election.

cmccabe replied:
Even in an unclean leader election, we don't want to change the leader unless we need to.

junrao reviewed (commented):
@cmccabe : Thanks for the new updates. A few more comments below.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,894 @@
+            throw new RuntimeException("Tried to create partition " + record.topicId() +
+                ":" + record.partitionId() + ", but no partition with that id was found.");
+        }
+        PartitionControlInfo newPartitionInfo = prevPartitionInfo.merge(record);
```
Hmm, merge bumps up the leaderEpoch. It seems that this needs to be persisted in the metadata log?

cmccabe replied:
The leader epoch is managed implicitly -- every time a PartitionChangeRecord appears, the epoch is bumped if the leader is not NO_LEADER_CHANGE.

cmccabe committed:
Some fixes and improvements

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -0,0 +1,908 @@
+                        setErrorCode(Errors.INVALID_REQUEST.code()));
+                    continue;
+                }
+                records.add(new ApiMessageAndVersion(new PartitionChangeRecord().
```
Hmm, we should set the leader to NO_LEADER_CHANGE, right?

cmccabe replied:
That is the default so we don't need to set it unless we're changing it

junrao replied:
We have NO_LEADER_CHANGE as the default for the serialized data. However, the active controller replays the PartitionChangeRecord created in memory, which defaults leader to NO_LEADER_CHANGE, right?

junrao commented on metadata/src/main/resources/common/metadata/PartitionRecord.json:
```diff
@@ -34,6 +34,8 @@
     { "name": "LeaderEpoch", "type": "int32", "versions": "0+", "default": "-1",
-      "about": "An epoch that gets incremented each time we change the ISR." }
+      "about": "An epoch that gets incremented each time we change the ISR." },
+    { "name": "PartitionEpoch", "type": "int32", "versions": "0+", "default": "-1",
```
Could you file a separate jira to follow up on PartitionEpoch post 2.8?

cmccabe replied:
filed KAFKA-12349

junrao commented on metadata/src/main/resources/common/metadata/PartitionRecord.json:
```diff
@@ -34,6 +34,8 @@
     { "name": "LeaderEpoch", "type": "int32", "versions": "0+", "default": "-1",
-      "about": "An epoch that gets incremented each time we change the ISR." }
+      "about": "An epoch that gets incremented each time we change the ISR." },
+    { "name": "PartitionEpoch", "type": "int32", "versions": "0+", "default": "-1",
```
Could you file a separate jira to follow up on PartitionEpoch post 2.8?

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. Just one more comment.

junrao reviewed (commented):
@cmccabe : Still one more comment.

junrao reviewed (approved):
@cmccabe : Thanks for the latest PR. LGTM. The last comment is no longer an issue after regenerating the code.

Overall, the new controller code is quite elegant and much easier to follow than the old controller code.

cmccabe commented:
Thanks, @junrao ! :)
